### ISSUE TITLE
Fine-tuning a pre-trained ASR model on a new dataset

### ISSUE THREAD

**Issue by danoneata** [espnet/espnet#2930]
Hello!

I'm trying to fine-tune a pre-trained Espnet2 ASR model on a new dataset. I was not able to find a recipe for this exact scenario (although there seem to be [some instructions available for TTS](https://github.com/espnet/espnet/tree/master/egs2/jvs/tts1)) and I would need some guidance. Concretely, I'm interested in fine-tuning [the LibriSpeech-trained Transformer model](https://zenodo.org/record/3966501#.YBlN1eCxXOQ) on a smaller audio dataset.

Here are some questions that I have:

1. Do I need to run both stage 9 (collect stats) and stage 10 (the effective training) from the [`egs2/TEMPLATE/asr1/asr.sh`](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh) script?
2. For either step, do I have to modify the configuration file (`config.yaml`) of the pre-trained model? For example, I would be tempted to remove the information that is not related to the model configuration (such as paths) and specify it separately.
3. For stage 10 (training) should I use the normalizing statistics (`feats_stats.npz`) computed on the original data (and which are packaged with the pre-trained model) or should I use the ones computed on the target dataset?
4. In general, what are some good practices for fine-tuning? For example, is better to freeze part of the network (_e.g._, the encoder) or to fine-tune the entire network? Is there any publish work on this topic?

Thank you!
– Dan

**Comment by sw005320** [comment]
Thanks for bringing interesting discussions!
I can answer some of them based on my experience.

> 1. Do I need to run both stage 9 (collect stats) and stage 10 (the effective training) from the [`egs2/TEMPLATE/asr1/asr.sh`](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh) script?

Yes. 
In general, the target dataset would have different statistics, and it would be safe to collect the in-domain statistics in stage 9.

> 3. For stage 10 (training) should I use the normalizing statistics (`feats_stats.npz`) computed on the original data (and which are packaged with the pre-trained model) or should I use the ones computed on the target dataset?

If you mix the original and target data, I recommend you get the mixed data statistics.
Otherwise, you can use that of the target data obtained in stage 9.

> 4. In general, what are some good practices for fine-tuning? For example, is it better to freeze part of the network (_e.g._, the encoder) or fine-tune the entire network? Is there any publish work on this topic?

I saw many papers, and some of them conclude that freezing the part of the network is better, but my conclusion is that fine-tuning the entire network is easy and robust in general. You'd get some primary benefits of fine-tuning in this setup. If you want to get further small improvement, you can try to freeze a part of the network (e.g., the lower layers in the language adaptation scenario and higher layers in the noise or other acoustic environment adaptation scenario).

**Comment by danoneata** [comment]
Thank you for the quick and detailed answer, Shinji!

>  In general, the target dataset would have different statistics, and it would be safe to collect the in-domain statistics in stage 9.

If I understand correctly, the `collect_stats` stage also computes the shapes of speech and text; we cannot run the training step (stage 10) without these quantities, can we?

> If you mix the original and target data, I recommend you get the mixed data statistics.

Just to double check: mixing the data statistics can be achieved by adding each of the three quantities (`count`, `sum`, `sum_squared`) corresponding to the two datasets:
```python
stats = {k: stats1[k] + stats2[k] for k in "count sum sum_squared".split()}
```

> I saw many papers, and some of them conclude that freezing the part of the network is better, but my conclusion is that fine-tuning the entire network is easy and robust in general. You'd get some primary benefits of fine-tuning in this setup. If you want to get further small improvement, you can try to freeze a part of the network (e.g., the lower layers in the language adaptation scenario and higher layers in the noise or other acoustic environment adaptation scenario).

Thanks a lot for sharing your thoughts on this! Would you suggest using a smaller learning rate (than the one used for pre-training) when performing the fine-tuning or it doesn't matter that much?

**Comment by sw005320** [comment]
> > In general, the target dataset would have different statistics, and it would be safe to collect the in-domain statistics in stage 9.
> 
> If I understand correctly, the `collect_stats` stage also computes the shapes of speech and text; we cannot run the training step (stage 10) without these quantities, can we?

Correct.

> > If you mix the original and target data, I recommend you get the mixed data statistics.
> 
> Just to double check: mixing the data statistics can be achieved by adding each of the three quantities (`count`, `sum`, `sum_squared`) corresponding to the two datasets:
> 
> ```python
> stats = {k: stats1[k] + stats2[k] for k in "count sum sum_squared".split()}
> ```

Yes.
But it may be easier to combine them in the data preparation stage and run it from the beginning.

> > I saw many papers, and some of them conclude that freezing the part of the network is better, but my conclusion is that fine-tuning the entire network is easy and robust in general. You'd get some primary benefits of fine-tuning in this setup. If you want to get further small improvement, you can try to freeze a part of the network (e.g., the lower layers in the language adaptation scenario and higher layers in the noise or other acoustic environment adaptation scenario).
> 
> Thanks a lot for sharing your thoughts on this! Would you suggest using a smaller learning rate (than the one used for pre-training) when performing the fine-tuning or it doesn't matter that much?

This is case-by-case.
I prefer to use the vanilla SGD with a small learning rate, which is very safe. 
But it often turns out that the standard optimization would work in many cases...

**Comment by danoneata** [comment]
Thank you for the further clarifications!

**Comment by elhamadjidi** [comment]
> Thank you for the further clarifications!

Hi Dan, 
I want to fine-tune any pre-trained model with children's voices, I want to make a TTS with children's voices. Can you tell me where to start? I am so confused with all the information here.

**Comment by danoneata** [comment]
@elhamadjidi I haven't kept up with the latest developments in ESPnet. From what I recall, a tutorial on how to finetune a pretrained TTS model is available [here](https://github.com/espnet/espnet/tree/master/egs2/jvs/tts1) (I'm not sure if there are other more up-to-date resources). A list of pretrained TTS models is available [here](https://zenodo.org/communities/espnet/search?page=1&size=20&q=&keywords=speech-synthesis). I don't think there are models trained on children's voices, but maybe you can start from models pretrained on common TTS datasets, such as LJSpeech or LibriSpeech. Hope this helps...

**Comment by elhamadjidi** [comment]
> @elhamadjidi I haven't kept up with the latest developments in ESPnet. From what I recall, a tutorial on how to finetune a pretrained TTS model is available [here](https://github.com/espnet/espnet/tree/master/egs2/jvs/tts1) (I'm not sure if there are other more up-to-date resources). A list of pretrained TTS models is available [here](https://zenodo.org/communities/espnet/search?page=1&size=20&q=&keywords=speech-synthesis). I don't think there are models trained on children's voices, but maybe you can start from models pretrained on common TTS datasets, such as LJSpeech or LibriSpeech. Hope this helps...

Yeah, it helped, Thank you. One more question, I tried to do 2-5 stages with LJSpeech with my own dataset and it took 11 hours in google Colab and then stopped due to Colab problems. Do you suggest any platform that can handle all these computations in a reasonable time? I have a MacBook Pro, with no GPUs.
